### ISSUE TITLE
Python import pyclblast can't find /usr/local/lib/libclblast.so.1

### ISSUE THREAD

**Issue by CaiusTSM** [CNugteren/CLBlast#412]
Hi, when I import pyclblast it can't find the `libclblast.so.1` file when installed with `pip install .`. I would prefer to not have to edit my LD_LIBRARY_PATH to use pyclblast and instead have it copy the shared objects to the site packages or wherever it needs to be so that it can be found by python by default.

**Comment by CNugteren** [comment]
I agree that that would be nice indeed :-) But at the moment it is designed that you need to both install CLBlast and `pyclblast` to make it work. I hope that that is clear from the [README](https://github.com/CNugteren/CLBlast/tree/master/src/pyclblast). If you can make it such that `pyclblast` installs CLBlast itself, feel free to make a PR, but I don't think that will be easy.

**Comment by CaiusTSM** [comment]
A quick fix for now is to add `runtime_library_dirs=["/usr/local/lib"]` to `setup.py`. This allows it to find the shared object if CLBlast is installed without the need to manually add it to `LD_LIBRARY_PATH`.
```
ext_modules = list()
ext_modules.append(
    Extension(
        "pyclblast",
        ["src/pyclblast.pyx"],
        libraries=["clblast"],
        runtime_library_dirs=["/usr/local/lib"],
        language="c++"
    )
)
```

**Comment by CNugteren** [comment]
OK, makes sense. I've implemented this in this PR, could you have a look if it makes sense to you?
https://github.com/CNugteren/CLBlast/pull/413

**Comment by CaiusTSM** [comment]
Yes it seems OK. This could also be done for windows. The location that cmake places the files is documented here: https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html

`/usr/local` for UNIX and `C:\Program Files\${Project_Name}` for windows. So on windows (and I need to check this), the dll (shared object) would be in `C:\Program Files\CLBlast\lib`.

I think I might put in the effort to make `setup.py` actually build and install CLBlast as well (windows would be required to have nmake in path / linux needs make). It may also be better to have the `setup.py` be in the top-most directory, but that depends on how PyPi works (where does it look for `setup.py`).

**Comment by CNugteren** [comment]
I've added the Windows case to the PR as well. I don't have a Windows system so can't check - but if it is wrong it can't hurt either I guess.

I rather not combine CLBlast and the Python package, because it makes everything a lot more complicated. By having it separate we can easily install them separately and have installation instructions (and troubleshooting) separately as well.

**Comment by CaiusTSM** [comment]
Ok, how about having pyclblast be its own github repository. Then CLBlast could be added as a git sub-module. This allows for the two projects to have their own github issues separated. It also makes it clear that they are two separate things where one requires the other to be installed. In addition, the pyclblast github page can have its own README, and API documentation. For the pypi package, the setup.py can check to see if CLBlast is already installed (by checking for the shared object / dll in the previously mentioned default locations), if it's not already installed, it can either give a warning/error, or build and install from source (`cmake ..`, `cmake --build .`, `cmake --install .` <- requires sudo). I have already modified my setup.py to build and install from source so I could add this functionality if desired. The reason for the whole build and install from source (if not already installed) thing is because if I create a library built on top of pyclblast (say a ML lib), then I would like the user to just be able to do `sudo pip install lib` and it will take care of everything (maximum simplicity, but if they need to configure CLBlast they can still build and install it themselves). 

I will boot up Windows and check where CLBlast installs to to confirm the runtime dir.

**Comment by CNugteren** [comment]
> Ok, how about having pyclblast be its own github repository

I don't really see how that would change anything or benefit, it just looks like work to me :-) The `pyclblast` folder already has its own README, so that is not really an advantage to me. But if you can convince me I can consider it.

> For the pypi package, the setup.py can check to see if CLBlast is already installed, if it's not already installed, it can either give a warning/error, or build and install from source.

I'm open to this, but I am afraid this will just lead to a lot of work and problems, also for users. You already mention one: what if they don't want to install as root but as a user or in a virtualenv? Maybe this problem and many others are easily solveable, but to me it looks a lot simpler for everyone if there are two things. OpenCL itself is anyway also not pre-installed, and similarly to the current `pyclblast` situation also not installed when you `pip install pyopencl`. And then you'll have to think of what happens when there are compilation errors? Incompatible compilers/cmake versions? Users might want to provide a specific flag to CMake? What if it is installed in a non-standard way. Users might want to use specific/older/newer versions, etc. I'm just putting a lot of questions out there, not to get answers but just to illustrate that there might be problems.

That said, if you want to try this out and come up with a reasonably nice solution, then feel free to open up a PR and I'll review it. Ideally this new feature should be optional as well, so that we don't bother those that already have CLBlast installed separately.

**Comment by CaiusTSM** [comment]
Ok, I will make a new issue for it. The build from source would be optional, the default `pip install pyclblast` would behave the same way it does now (the more simple option as the default). Building from source would be with something like `pip install pyclblast --install-option="--install-clblast"`.


I built and installed CLBlast on Windows 10 64 bit (with visual studio 2019) and the opencl light sdk: https://github.com/GPUOpen-LibrariesAndSDKs/OCL-SDK/releases/tag/1.0. The environment variable `OPENCL_ROOT` had to be created and set to `C:\OCL_SDK_Light` for cmake to find it.

CLBlast was built with these commands:
```
mkdir build
cd build
cmake .. -A x64
cmake --build . --config Release
```

And installed with `cmake --install .` (Note: the console was run as admin).

Here are the default install directories/output:

```
-- Install configuration: "Release"
-- Installing: C:/Program Files (x86)/clblast/lib/clblast.lib
-- Installing: C:/Program Files (x86)/clblast/lib/clblast.dll
-- Installing: C:/Program Files (x86)/clblast/include/clblast_half.h
-- Installing: C:/Program Files (x86)/clblast/include/clblast.h
-- Installing: C:/Program Files (x86)/clblast/include/clblast_c.h
-- Installing: C:/Program Files (x86)/clblast/lib/cmake/CLBLast/CLBlastConfig.cmake
-- Installing: C:/Program Files (x86)/clblast/lib/cmake/CLBLast/CLBlastConfig-release.cmake
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_copy_fast.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_copy_pad.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_transpose_fast.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_transpose_pad.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xaxpy.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xdot.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xger.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xgemm.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xgemm_direct.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xgemv.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_invert.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_xconvgemm.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_routine_xgemm.exe
-- Installing: C:/Program Files (x86)/clblast/bin/clblast_tuner_routine_xtrsv.exe
```

To be safe you could probably add both `C:\Program Files (x86)\clblast\lib` and `C:\Program Files\clblast\lib` to the runtime dirs.

**Comment by CNugteren** [comment]
> The build from source would be optional, the default pip install pyclblast would behave the same way it does now (the more simple option as the default). Building from source would be with something like pip install pyclblast --install-option="--install-clblast".

Sounds like a good idea indeed. Let me know if you have any questions about the current build system.

> To be safe you could probably add both C:\Program Files (x86)\clblast\lib and C:\Program Files\clblast\lib to the runtime dirs.

I've just done that in the PR and merged it in. Thanks!

**Comment by CaiusTSM** [comment]
I just looked at the commit and noticed that you used `CLBlast` in the path which is slightly wrong. It's case sensitive and seems to create the all lower case directory `clblast`.

**Comment by CNugteren** [comment]
Thanks for checking, I've made a new PR to fix that.
